Improve extension interface

mfeurer opened the pull request:
Continuation of #647:

* [x] Remove obtain_arff_trace and return the trace headers in each fold and later on check that they're all the same.
* [x] Simplify run_model_on_fold to accept X and y and do not split the data itself
* [x] Simplify the return value of run_model_on_fold
* ~~[ ] Have functionality for transforming predictions into arff lines in the different task classes.~~ (As discussed with JvR offline, we'll keep task-specific code here until the maintenance becomes an issue as for now the code would be more complex if such transformations would be in the task classes).
* ~~[ ] Figure out how to best handle dependencies (as raised here)~~ (this is beyond the scope of this PR)
* [x] add more unit tests

janvanrijn commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1029,11 +1031,15 @@ def _run_model_on_fold(
         task: 'OpenMLTask',
+        X_train: Union[np.ndarray, scipy.sparse.spmatrix, pd.DataFrame],
+        y_train: np.ndarray,
         rep_no: int,
```
can be removed, together with `fold_no`, `sample_no`

mfeurer replied:
`rep_no` and `fold_no` are necessary to create trace objects.

janvanrijn commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1029,11 +1031,15 @@ def _run_model_on_fold(
         rep_no: int,
         fold_no: int,
         sample_no: int,
         add_local_measures: bool,
```
can be removed as it can be handled a level up

mfeurer replied:
Done

PGijsbers commented on openml/extensions/extension_interface.py:
```diff
@@ -1,12 +1,15 @@
     from openml.flows import OpenMLFlow
     from openml.tasks.task import OpenMLTask
-    from openml.runs.trace import OpenMLRunTrace, OpenMLTraceIteration
+    from openml.runs.trace import OpenMLRunTrace, OpenMLTraceIteration  # noqa F401
```
Why is OpenMLRunTrace even imported in the first place? It does not seem to be used in the file?

mfeurer replied:
Fixed, it's now used.

PGijsbers commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1119,17 +1123,12 @@ def _run_model_on_fold(
-            Determines whether to calculate a set of measures (i.e., predictive accuracy)
-            locally,
-            to later verify server behaviour.
 
```
Looks like `y_train`, `X_test` and `classes` are missing from docstring?

mfeurer replied:
Thanks, I forgot to copy them from the parent class.

PGijsbers commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1268,154 +1249,50 @@ def _prediction_to_probabilities(
+                # Remap the probabilities in case there was a class missing at training time
+                # By default, the classification targets are mapped to be zero-based indices to the
+                # actual classes. Therefore, the model_classes contain the correct indices to the
+                # correct probability array (the actualy array might be incorrect if there are some
```
typo: actualy

mfeurer replied:
Just corrected this.

PGijsbers commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1268,154 +1249,50 @@ def _prediction_to_probabilities(
+                # correct probability array (the actualy array might be incorrect if there are some
+                # classes not present during train time).
+                proba_y_new = np.zeros((proba_y.shape[0], len(classes)))
+                for idx, model_class in enumerate(model_classes):
```
I don't understand what you're trying to do here.
If targets are mapped to be zero-based indices, wouldn't `enumerate(model_classes)` just evaluate to `(1,1), (2,2), ..., (Kt, Kt)` where `Kt` is the number of classes in the training data?
If so why not do either
```
proba_y_new = np.zeros((proba_y.shape[0], len(classes)))
proba_y_new[:proba_y.shape[0], :proba_y.shape[1]] = proba_y
```
or use `np.hstack` to pad zero-columns?

mfeurer replied:
The zero-columns might need to be in the middle of the array. I added an example to the comment.

PGijsbers commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1268,154 +1249,50 @@ def _prediction_to_probabilities(
-                warnings.warn(
-                    "Repeat %d Fold %d: estimator only predicted for %d/%d classes!"
-                    % (rep_no, fold_no, proba_y.shape[1], len(task.class_labels))
+                message = "Estimator only predicted for {}/{} classes!".format(
```
in what scenario should be expect `len(classes)` to differ from `len(task.class_labels)`? (Also, what even is the difference between `classes` and `task.class_labels`?)

mfeurer replied:
There's no difference, thanks for spotting this issue, it'll make the API a bit easier.

PGijsbers commented on openml/extensions/sklearn/extension.py:
```diff
@@ -1268,154 +1249,50 @@ def _prediction_to_probabilities(
-
         if isinstance(task, (OpenMLClassificationTask, OpenMLLearningCurveTask)):
 
+            if classes is None:
```
I would personally prefer argument checking to be done at the start of the method call such that all constraints are clear, and `TypeError` lead to a quick fail (not e.g. after running a whole grid search!). Even though I know it introduces an extra line of code :)

mfeurer replied:
That check will go away when using `task.class_labels`. I will add a check that `X_test` and `y_train` are given at the top of the function.

PGijsbers commented on openml/runs/functions.py:
```diff
@@ -391,26 +395,103 @@ def _run_task_get_arffcontent(
+    classes = None
 
+    n_fit = 0
     for rep_no in range(num_reps):
```
maybe replace this with `itertools.product(num_reps, num_folds, num_samples)` so we get rid of two levels of indentation and can maybe spare some ugly line-breaks below.

mfeurer replied:
Done

PGijsbers commented on openml/runs/functions.py:
```diff
@@ -391,26 +395,103 @@ def _run_task_get_arffcontent(
     for rep_no in range(num_reps):
         for fold_no in range(num_folds):
             for sample_no in range(num_samples):
+                n_fit += 1
```
I don't see a reason not to use `enumerate` (it looks like you want to start at 1, which is done with `enumerate(Sequence, start=1)`

mfeurer replied:
Done, thanks for the tip with `start=1`.

PGijsbers commented on openml/runs/functions.py:
```diff
@@ -391,26 +395,103 @@ def _run_task_get_arffcontent(
+                else:
+                    raise TypeError(type(task))
+
                 arff_datacontent.extend(arff_datacontent_fold)
```
What is the use `arff_datacontent_fold`? It is defined in line 446 but I don't see it being used anywhere else. Is this an error? Or is it legacy? If it is legacy, I would prefer `arff_datacontent.extend([])  # legacy` or something equivalent.

mfeurer replied:
It's not used any more so I removed it.

PGijsbers commented on openml/runs/functions.py:
```diff
@@ -432,10 +513,13 @@ def _run_task_get_arffcontent(
+    if len(traces) > 0:
+        if len(traces) != n_fit:
+            raise ValueError(
+                'Did not find enough traces (expected %d, found %d)' % (n_fit, len(traces))
```
`.format`

mfeurer replied:
Done

PGijsbers commented on openml/runs/trace.py:
```diff
@@ -346,9 +348,23 @@ def trace_from_xml(cls, xml):
     def __str__(self):
         return '[Run id: %d, %d trace iterations]' % (
-            self.run_id,
+            -1 if self.run_id is None else self.run_id,
```
`.format`

mfeurer replied:
Done

PGijsbers commented on openml/runs/trace.py:
```diff
@@ -448,3 +464,14 @@ def __str__(self):
+        if not isinstance(other, OpenMLTraceIteration):
+            return False
+        attributes = [
+            'repeat', 'fold', 'iteration', 'setup_string', 'evaluation', 'selected', 'paramaters',
```
I noticed the typo `paramaters` here, and also noticed this is present all throughout the init method. Could you fix that up there too?
Also, it looks like this code probably isn't tested then? Since `getattr` should raise an exception as this object does not have an attribute `paramater` as (somehow) the attribute to the object itself _is_ spelled correctly :)

mfeurer replied:
I fixed the attribute and removed the equals function as it was not used.

PGijsbers commented on openml/tasks/task.py:
```diff
@@ -177,6 +177,20 @@ def __init__(self, task_id, task_type_id, task_type, data_set_id,
 
+    def get_X(self, dataset_format='array'):
+        """Get data associated with the current task.
+
```
type hint, doc string with legal values for `dataset_format` parameter.

mfeurer replied:
Done.

PGijsbers commented on openml/tasks/task.py:
```diff
@@ -108,7 +108,7 @@ def __init__(self, task_id, task_type_id, task_type, data_set_id,
 
-    def get_X_and_y(self):
+    def get_X_and_y(self, dataset_format='array'):
         """Get data associated with the current task.
```
type hint, doc string with legal values for `dataset_format` parameter.

mfeurer replied:
Done.

PGijsbers commented on openml/runs/trace.py:
```diff
@@ -346,9 +348,23 @@ def trace_from_xml(cls, xml):
         return cls(run_id, trace)
 
+    @classmethod
+    def merge_traces(cls, traces: List['OpenMLRunTrace']):
```
return type not specified in typehint

mfeurer replied:
Done.

PGijsbers commented on tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py:
```diff
@@ -1254,101 +1273,259 @@ def test_seed_model_raises(self):
-        self.assertEqual(len(arff_tracecontent), 0)
+        self.assertIsNone(trace)
+
+        self._check_fold_timing_evaluations(fold_evaluations, num_repeats, num_folds,
```
since neither `num_repeats` nor `num_folds` seem to be used throughout the test other than here, I would prefer the call to just be
```
self._check_fold_timing_evaluations(fold_evaluations, 
    num_repeats=1, num_folds=1,
    task_type=task.task_type_id, check_scores=False)
```
which makes that clear.

mfeurer replied:
Done.

PGijsbers commented on tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py:
```diff
@@ -1254,101 +1273,259 @@ def test_seed_model_raises(self):
+        self.assertIsInstance(trace, OpenMLRunTrace)
+        self.assertEqual(len(trace.trace_iterations), 2)
+
         self._check_fold_timing_evaluations(fold_evaluations, num_repeats, num_folds,
```
same note on `num_repeats` and `num_folds`

mfeurer replied:
Done.

PGijsbers commented on tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py:
```diff
@@ -1254,101 +1273,259 @@ def test_seed_model_raises(self):
+                rep_no=0,
+                classes=task.class_labels,
+            )
+            pred_2, proba_2, _, _ = self.extension._run_model_on_fold(
```
`proba_1` and `proba_2` are unused? so no need to name them either.
You can ignore any amount of return arguments with `pred_1, *_`.

mfeurer replied:
I added some additional asserts to actually use them.

PGijsbers commented on tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py:
```diff
@@ -1254,101 +1273,259 @@ def test_seed_model_raises(self):
+        # trace. SGD does not produce any
+        self.assertIsNone(trace)
+
+        self._check_fold_timing_evaluations(fold_evaluations, num_repeats, num_folds,
```
`num_repeats`

mfeurer replied:
Done.

PGijsbers commented on tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py:
```diff
@@ -1254,101 +1273,259 @@ def test_seed_model_raises(self):
+        # trace. SGD does not produce any
+        self.assertIsNone(trace)
+
+        self._check_fold_timing_evaluations(fold_evaluations, num_repeats, num_folds,
```
`num_repeats`

mfeurer replied:
Done.

mfeurer commented on openml/runs/trace.py:
```diff
@@ -346,9 +348,35 @@ def trace_from_xml(cls, xml):
+                merged_trace[key] = iteration
+                previous_iteration = key
+
+        return cls(None, merged_trace)
```
Thanks for the suggestion. I'm not 100% sure whether this will allow the same flexibility, for example, the first statement would only check the iterations in one trace, and it would be hard to extract a useful error message out of it. Therefore, i suggest leaving the code as it is.

PGijsbers replied:
Right, I missed that.
```python
all_iterations = [iter for trace in traces for iter in trace]
for (iter1, iter2) in zip(all_iterations, all_iterations[1:]):
   if sorted(iter1.parameters) != sorted(iter2.parameters):
       raise ValueError(...)

```
should then work fine? But I am fine leaving this as is (and perhaps refactoring it later if we can find something we all like).

mfeurer commented:
Okay @PGijsbers I found the issue. After some offline discussion with @janvanrijn I figured out that we can see the reason why there is no trace in the error log of the run. It turns out that there were three bugs:

1. there was an indexing error which caused the predictions to have wrong indices and therefore make the evaluation fail. I fixed this.
2. in case there was no evaluation, after 200 seconds the test continued and raised the cryptic error message that there was no trace associated. I added a useful error message.
3. In case there were no evaluations, due to recently added run caching, the test could not retrieve any new evaluations by the server because it would continue to check the cached run. I added a flag to ignore the cached run and download it again from the server.

All in all, this should be good to merge now.
